File: 2022/02/rust/src/main.rs
use std::io;

#[derive(PartialEq, Eq, Clone, Copy)]
enum Shape {
    Rock,
    Paper,
    Scissors,
}

impl Shape {
    fn score(&self) -> i32 {
        match *self {
            Shape::Rock => 1,
            Shape::Paper => 2,
            Shape::Scissors => 3,
        }
    }

    fn wins(&self) -> Shape {
        match *self {
            Shape::Rock => Shape::Scissors,
            Shape::Paper => Shape::Rock,
            Shape::Scissors => Shape::Paper,
        }
    }

    fn loses(&self) -> Shape {
        match *self {
            Shape::Rock => Shape::Paper,
            Shape::Paper => Shape::Scissors,
            Shape::Scissors => Shape::Rock,
        }
    }

    fn draws(&self) -> Shape {
        *self
    }
}

enum Outcome {
    Lose,
    Draw,
    Win,
}

impl Outcome {
    fn score(&self) -> i32 {
        match *self {
            Outcome::Lose => 0,
            Outcome::Draw => 3,
            Outcome::Win => 6,
        }
    }
}



fn main() {
    let mut score = 0;
    for line in io::stdin().lines() {
        let op_move = line.as_ref().unwrap().chars().next().unwrap();
        let outcome_code = line.as_ref().unwrap().chars().nth(2).unwrap();

        let op_shape = match op_move {
            'A' => Shape::Rock,
            'B' => Shape::Paper,
            'C' => Shape::Scissors,
            _ => panic!("Unknown move"),
        };
        let outcome = match outcome_code {
            'X' => Outcome::Lose,
            'Y' => Outcome::Draw,
            'Z' => Outcome::Win,
            _ => panic!("Unknown outcome"),
        };
        let my_shape = match outcome {
            Outcome::Lose => op_shape.wins(),
            Outcome::Draw => op_shape.draws(),
            Outcome::Win => op_shape.loses(),
        };
        score += my_shape.score() + outcome.score();
    }
    println!("{score}");
}

File: 2022/02/python/main.py
import sys

ROCK = 0
PAPER = 1
SCISSORS = 2

score = 0
for row in sys.stdin:
    op_move, outcome_code = row.strip().split(" ")

    if op_move == "A":
        op_shape = ROCK
    elif op_move == "B":
        op_shape = PAPER
    elif op_move == "C":
        op_shape = SCISSORS

    if outcome_code == "X":
        my_shape = (op_shape - 1) % 3
    elif outcome_code == "Y":
        my_shape = op_shape
    elif outcome_code == "Z":
        my_shape = (op_shape + 1) % 3

    score += my_shape + 1
    outcome = (my_shape - op_shape) % 3
    if outcome == 1:
        score += 6
    elif outcome == 0:
        score += 3
    elif outcome == 2:
        score += 0

print(score)
